pkg/operator/operator.go

```diff
@@ -269,31 +274,82 @@ func (o *Operator) setupAdmissionWebhooks(ctx context.Context) error {
 
 	// Keep setting the caBundle in the expected webhook configurations.
-	go func() {
-		// Only inject if we've an explicit CA bundle ourselves. Otherwise the webhook configs
-		// may already have been created with one.
-		if len(caBundle) == 0 {
-			return
-		}
-		// Initial sleep for the client to initialize before our first calls.
-		// Ideally we could explicitly wait for it.
-		time.Sleep(5 * time.Second)
+	err = ctrl.NewControllerManagedBy(o.manager).
+		Named("validating-webhook-cert-updater").
+		For(
+			&arv1.ValidatingWebhookConfiguration{},
+			builder.WithPredicates(predicate.Funcs{
+				CreateFunc: func(ev event.CreateEvent) bool {
+					return ev.Object.GetName() == webhookConfigName(o.opts)
+				},
+				UpdateFunc: func(ev event.UpdateEvent) bool {
+					return ev.ObjectNew.GetName() == webhookConfigName(o.opts)
+				},
+				GenericFunc: func(ev event.GenericEvent) bool {
+					return ev.Object.GetName() == webhookConfigName(o.opts)
+				},
+			}),
+		).
+		Complete(reconcile.Func(func(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+			logger, _ := logr.FromContext(ctx)
+			logger.WithValues("operator", req.NamespacedName).Info("reconciling validatingwebhookconfiguration")
 
-		for {
 			if err := o.setValidatingWebhookCABundle(ctx, caBundle); err != nil {
-				o.logger.Error(err, "Setting CA bundle for ValidatingWebhookConfiguration failed")
+				return reconcile.Result{}, err
 			}
+			return reconcile.Result{}, nil
+		}))
+	if err != nil {
+		return errors.Wrap(err, "create validating webhook updater controller")
+	}
+
+	err = ctrl.NewControllerManagedBy(o.manager).
+		Named("mutating-webhook-cert-updater").
+		For(
+			&arv1.MutatingWebhookConfiguration{},
+			builder.WithPredicates(predicate.Funcs{
+				CreateFunc: func(ev event.CreateEvent) bool {
+					return ev.Object.GetName() == webhookConfigName(o.opts)
+				},
+				UpdateFunc: func(ev event.UpdateEvent) bool {
+					return ev.ObjectNew.GetName() == webhookConfigName(o.opts)
+				},
+				GenericFunc: func(ev event.GenericEvent) bool {
+					return ev.Object.GetName() == webhookConfigName(o.opts)
+				},
+			}),
+		).
+		Complete(reconcile.Func(func(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+			logger, _ := logr.FromContext(ctx)
+			logger.WithValues("operator", req.NamespacedName).Info("reconciling mutatingwebhookconfiguration")
+
 			if err := o.setMutatingWebhookCABundle(ctx, caBundle); err != nil {
-				o.logger.Error(err, "Setting CA bundle for MutatingWebhookConfiguration failed")
-			}
-			select {
-			case <-ctx.Done():
-				return
-			case <-time.After(time.Minute):
+				return reconcile.Result{}, err
 			}
+			return reconcile.Result{}, nil
+		}))
+	if err != nil {
+		return errors.Wrap(err, "create mutating webhook updater controller")
+	}
+
+	// This method runs after the webhook server and client are started but
+	// before the controllers are started. Running it here prevents race
+	// conditions.
```

Comment:
Can you elaborate more on why this is necessary? What are the specific race conditions you're concerned with?